pep-0484.txt

```diff
@@ -1120,6 +1120,56 @@ subtype of ``Type[Base]``::
       ...
 
 
+Annotating instance and class methods
+-------------------------------------
+
+In most cases the first argument of class and instance methods
+does not need to be annotated, and it is assumed to have the
+type of the containing class for instance methods, and a type object
+type corresponding to the containing class object for class methods.
+In addition, the first argument in an instance method can be annotated
+with a type variable. In this case the return type uses the same
+type variable, thus making that method a generic function. For example::
+
+  T = TypeVar('T', bound='Copyable')
+  class Copyable:
+      def copy(self: T) -> T:
+          # return a copy of self
+
+  class C(Copyable): ...
+  c = C()
+  c2 = c.copy()  # type here should be C
+
+The same applies to class methods using ``Type[]`` in an annotation
+of the first argument::
+
+  T = TypeVar('T', bound='C')
+  class C:
+      def factory(cls: Type[T]) -> T:
+          # make a new instance of cls
+
+  class D(C): ...
+  d = D.factory()  # type here should be D
+
+Note that some type checkers may apply restrictions on this use, such as
+requiring an appropriate upper bound for the type variable used
+(see examples). Also this use case should not be confused with the use
```

Comment:
@gvanrossum Anyway, I don't want to put it back again, only if you think it will help. Here is version without ``__add__``
```python
T = TypeVar('T')
class Num(Generic[T]):
    def __init__(self, val: T) -> None: ...
    def compare(self, other: 'Num[T]') -> bool: ...
Num(1).compare(Num(3.14)) # Rejected by type checker

N = TypeVar('N', bound='Float')
class Float:
    def __init__(self, val) -> None: ...
    def compare(self: N, other: N) -> bool: ...
class Int(Float): ...
Int(1).compare(Float(3.14)) # OK, the inferred value of N is Float
```
Again, this works as expected in mypy.